src/inline.rs

```diff
@@ -375,41 +374,108 @@ impl<'s> Parser<'s> {
                 non_whitespace_encountered: false,
                 non_whitespace_last: None,
             });
+            self.attributes = None;
             self.push(EventKind::Enter(ty))
         }
     }
 
     fn parse_attributes(&mut self, first: &lex::Token) -> Option<ControlFlow> {
         if first.kind == lex::Kind::Open(Delimiter::Brace) {
-            let mut ahead = self.input.lexer.ahead().chars();
-            let (mut attr_len, mut has_attr) = attr::valid(std::iter::once('{').chain(&mut ahead));
-            attr_len = attr_len.saturating_sub(1); // rm {
-            if attr_len > 0 {
-                while attr_len > 0 {
-                    self.input.span = self.input.span.extend(attr_len);
-                    self.input.lexer = lex::Lexer::new(ahead.as_str());
-
-                    let (l, non_empty) = attr::valid(&mut ahead);
-                    attr_len = l;
-                    has_attr |= non_empty;
+            let elem_ty = self
+                .attributes
+                .take()
+                .unwrap_or(AttributesElementType::Word);
+            self.ahead_attributes(elem_ty, true)
+        } else {
+            None
+        }
+    }
+
+    fn ahead_attributes(
+        &mut self,
+        elem_ty: AttributesElementType,
+        opener_eaten: bool,
+    ) -> Option<ControlFlow> {
+        let start_attr = self.input.span.end() - usize::from(opener_eaten);
+        debug_assert!(self.input.src[start_attr..].starts_with('{'));
+
+        let mut line_start = start_attr;
+        let mut line_end = self.input.span_line.end();
+        let mut line = 0;
+        let mut end_attr = start_attr;
+        let mut attrs = attr::Attributes::new();
+        let mut parser = attr::Parser::new(&mut attrs, &self.input.src[start_attr..line_end]);
```

Comment:
We always restart parsing the attributes for each new line we receive here, yielding O^2 in terms of line count. Should be possible to keep the state in the inline::Parser and resume when next line has been retrieved.

Reply:
fixed, seemed to have helped some:

```
commit                                                           iai all    crit inline_attrs (MiB/s)
7af6f6efe clippy: allow blocks_in_if_conditions everywhere       8778060    52.361
659ab1e57 inline: take str per line instead of full inline iter  7783569    54.755
0897be0fd inline: add ControlFlow enum                           6625241    51.895
5c4e87b17 inline: parse multiline attributes                     7192831    25.255
285cef736 inline: resume attr parsing when new lines received    6928681    48.074
```